apps/emqx_bridge_azure_blob_storage/src/emqx_bridge_azure_blob_storage_connector.erl

```diff
@@ -0,0 +1,763 @@
+
+do_list_blobs(Worker, Container) ->
+    case erlazure:list_blobs(Worker, Container, [], infinity) of
+        {error, _} ->
+            error;
+        {L, _} when is_list(L) ->
+            ok
+    end.
+
+%%------------------------------------------------------------------------------
+%% `emqx_connector_aggreg_delivery' API
+%%------------------------------------------------------------------------------
+
+-spec init_transfer_state(emqx_connector_aggreg_delivery:buffer_map(), transfer_opts()) ->
+    transfer_state().
+init_transfer_state(BufferMap, Opts) ->
+    #{
+        upload_options := #{
+            action := ActionName,
+            blob := BlobTemplate,
+            container := Container,
+            max_block_size := MaxBlockSize,
+            pool := Pool
+        }
+    } = Opts,
+    Blob = mk_blob_name_key(BufferMap, ActionName, BlobTemplate),
+    #{
+        blob => Blob,
+        buffer => [],
+        buffer_size => 0,
+        container => Container,
+        max_block_size => MaxBlockSize,
+        num_blocks => 0,
+        pool => Pool,
+        started => false
+    }.
+
+mk_blob_name_key(BufferMap, ActionName, BlobTemplate) ->
+    emqx_template:render_strict(BlobTemplate, {?MODULE, {ActionName, BufferMap}}).
+
+-spec process_append(iodata(), transfer_state()) ->
+    transfer_state().
+process_append(IOData, TransferState0) ->
+    #{
+        buffer := Buffer,
+        buffer_size := BufferSize0
+    } = TransferState0,
+    TransferState0#{
+        buffer := [Buffer, IOData],
+        buffer_size := BufferSize0 + iolist_size(IOData)
+    }.
+
+-spec process_write(transfer_state()) ->
+    {ok, transfer_state()} | {error, term()}.
+process_write(TransferState0 = #{started := false}) ->
+    #{
+        pool := Pool,
+        blob := Blob,
+        container := Container
+    } = TransferState0,
+    %% TODO
+    %% Possible optimization: if the whole buffer fits the 5000 MiB `put_block_blob'
+    %% limit, we could upload the whole thing here.
+    case create_block_blob(Pool, Container, Blob) of
+        {ok, _} ->
+            TransferState = TransferState0#{started := true},
+            process_write(TransferState);
+        {error, Reason} ->
+            {error, Reason}
+    end;
+process_write(TransferState = #{started := true, buffer_size := 0}) ->
+    {ok, TransferState};
+process_write(TransferState0 = #{started := true}) ->
+    #{
+        buffer := Buffer,
+        buffer_size := BufferSize,
+        max_block_size := MaxBlockSize
+    } = TransferState0,
+    case BufferSize > MaxBlockSize of
+        true ->
+            {IOData, NewBuffer} = take_chunk(Buffer, MaxBlockSize),
+            ?tp(azure_blob_storage_will_write_chunk, #{}),
+            do_process_write(IOData, NewBuffer, TransferState0);
+        false ->
+            NewBuffer = [],
+            do_process_write(Buffer, NewBuffer, TransferState0)
+    end.
```

Comment:
Apparently having this append / write split backfires slightly. 🫠

Honestly both of those things feel suboptimal:
1. Call Azure API on every append, which is usually on the scale of 1-10s of KiBs assuming moderate event rate.
2. Cut >4GiB large buffer somewhere close to its end, which will likely cause 2-3x of that in heap memory consumption. This most likely will never happen, but still.

I wonder what if:
1. Introduce `min_block_size` config parameter, with something like 10 MiB as the default.
2. During append: append as usual, but put the buffer under some `next_block` map entry if its size is larger that `min_block_size` (i.e. minimize latency impact / overhead / cost (?) of too frequent API requests).
3. However, if it's to become larger than `max_block_size` does not append but instead put existing buffer under `next_block` key (i.e. avoid cutting iolists).
4. During write, write `next_block` if it's defined.

On more thing: currently, having a single append larger than 4 GiB is impossible I think, because of [this](https://github.com/emqx/emqx/blob/d61f4078e2b1ecd9493d8469de8057bf24307b31/apps/emqx_connector_aggregator/src/emqx_connector_aggreg_buffer.erl#L50). Admittedly, it's sort of an implicitly defined limit, but I'd argue still could be relied upon to get rid of complexity of "cut off exactly this much of this `iolist()`".